cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.services;
+
+import co.cask.cdap.api.artifact.ArtifactId;
+import co.cask.cdap.api.artifact.ArtifactScope;
+import co.cask.cdap.api.artifact.ArtifactVersion;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.internal.app.store.RunRecordMeta;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.ProgramStatus;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+
+/**
+ * ProgramLifecycleService tests.
+ */
+public class ProgramLifecycleServiceTest {
+
+  @Test
+  public void testEmptyRunsIsStopped() {
+    Assert.assertEquals(ProgramStatus.STOPPED, ProgramLifecycleService.getProgramStatus(Collections.emptyList()));
+  }
+
+  @Test
+  public void testProgramStatusFromSingleRun() {
+    RunRecordMeta record = RunRecordMeta.builder()
+      .setProgramRunId(NamespaceId.DEFAULT.app("app").mr("mr").run(RunIds.generate()))
+      .setStartTime(System.currentTimeMillis())
+      .setArtifactId(new ArtifactId("r", new ArtifactVersion("1.0"), ArtifactScope.USER))
+      .setStatus(ProgramRunStatus.PENDING)
+      .setSourceId(new byte[] { 0 })
+      .build();
+
+    // pending or starting -> starting
+    ProgramStatus status = ProgramLifecycleService.getProgramStatus(Collections.singleton(record));
+    Assert.assertEquals(ProgramStatus.STARTING, status);
+
+    record = RunRecordMeta.builder(record).setStatus(ProgramRunStatus.STARTING).build();
+    status = ProgramLifecycleService.getProgramStatus(Collections.singleton(record));
+    Assert.assertEquals(ProgramStatus.STARTING, status);
+
+    // running -> running
+    record = RunRecordMeta.builder(record).setStatus(ProgramRunStatus.RUNNING).build();
+    status = ProgramLifecycleService.getProgramStatus(Collections.singleton(record));
+    Assert.assertEquals(ProgramStatus.RUNNING, status);
+
+    // failed, killed, completed -> stopped
```

Comment:
since the method is called like `getProgramStatus(store.getActiveRuns(programId).values())`, there shouldn't be any record with failed, killed, completed statuses. Better to comment that these cases shouldn't happen but just testing an edge case. The only possible status to cause this method to return `STOPPED` is `SUSPENDED`

Reply:
that's the point of the test, to check that it returns the correct things

Reply:
right, I mean better to clarify these statuses shouldn't occur in real usage, and add a test case for suspended -> stopped

Reply:
I don't think suspend is actually implemented for anything. The old logic certainly did not handle those run statuses. I can add that, but suspended and resuming should not be stopped, they should be running.

I also disagree about adding a comment that this method normally gets only active run records, as the method does not and should not care what type of run records it gets. Instead, I can rename 'activeRunRecords' to just 'runRecords', since that is more clear.

Reply:
Ok a workflow can be suspended, which wasn't immediately clear to me. That means the previous logic was buggy, as it would say a suspended or resuming program was stopped, and disallow stopping that run.